op-node/rollup/derive/frame_queue.go

```diff
@@ -56,6 +61,68 @@ func (fq *FrameQueue) NextFrame(ctx context.Context) (Frame, error) {
 	return ret, nil
 }
 
+func (fq *FrameQueue) loadNextFrames(ctx context.Context) error {
+	// Find more frames if we need to
+	if len(fq.frames) == 0 {
+		data, err := fq.prev.NextData(ctx)
+		if err != nil {
+			return err
+		}
+
+		if frame, err := ParseFrames(data); err == nil {
+			fq.frames = append(fq.frames, frame...)
+		} else {
+			fq.log.Warn("Failed to parse frames", "origin", fq.prev.Origin(), "err", err)
+		}
+	}
+	return nil
+}
+
+func (fq *FrameQueue) prune() {
+	fq.frames = pruneFrameQueue(fq.frames)
+}
+
+// pruneFrameQueue prunes the frame queue to only hold contiguous and ordered
+// frames, conforming to Holocene frame queue rules.
+func pruneFrameQueue(frames []Frame) []Frame {
```

Comment:
It seems `pruneFrameQueue` will only prune within the scope of a single batch of frames returned by `prev.NextData`, in other words: it will not work across the batch borders?

Reply:
The frames from each batch are added to `fq.frames` so any frames from prior batches will be left in `fq.frames` and considered on the next run of pruning if they haven't yet formed a complete channel that can be processed.

Reply:
This assumption seems to be not satisfied since subsequent batches are only loaded when the current batch is exhaust: [source](https://github.com/ethereum-optimism/optimism/blob/5c17bf1e9e815a62eefe5d3678f790a597c9c805/op-node/rollup/derive/frame_queue.go#L66)

Reply:
We only need to prune the frame queue after we've _added_ frames to it, not after we've pulled some out into the next stage. And as Adrian correctly says, pruning happens to the whole queue, not only to the newly loaded ones.

But in fact, as you say correctly, the next stage would actually first pull out all frames before new frames are loaded. So we could actually set `fq.frames` to the frames that were loaded, instead of appending. At any rate, this behavior is fine because the next stage (channel bank) will also discard out of order frames. This is being implemented now.

I've actually refactored the code to only prune after loading new frames into the queue.

Reply:
Understood, so this PR only includes changes for `FrameQueue` but not other components like `ChannelBank`?

Reply:
yep, we're adding holocene derivation changes iteratively :)